### PR TITLE
minor: update checks(related to checkstyle issue#10100)

### PR DESCRIPTION
**Pull request by nmancus1**
Related to https://github.com/checkstyle/checkstyle/pull/10104

### Original reports from Checkstyle PR showing regression:

https://nmancus1.github.io/issue-10100_check_diff_reports_2021_06_11/diff_sevntu-check-regression_part_1/index.html

`NoMainMethodInAbstractClassCheck`- new NPE
`ForbidWildcardAsReturnType` - lots of new legit violations, we were missing violations on methods like `public Class<?>[] getClasses() {}`
`InnerClass` - symmetrical column changes
`StaticMethodCandidate` - new violations, we were missing violations previously on methods with `String` return type

-----------------------

https://nmancus1.github.io/issue-10100_check_diff_reports_2021_06_11/diff_sevntu-check-regression_part_2/index.html

`MoveVariableInsideIf` - symmetrical column number changes
`SimpleAccessorNameNotation` - symmetrical column number changes
`CustomDeclarationOrder` - symmetrical column number changes

---------------------------

### New reports showing no regression:

https://nmancus1.github.io/issue-10100_check_diff_reports_2021_06_16/diff_sevntu-check-regression_part_1/index.html

Symmetric column changes and new legitimate violations
`NoMainMethodInAbstractClassCheck`- NPE is gone

-----------------------------

https://nmancus1.github.io/issue-10100_check_diff_reports_2021_06_16/diff_sevntu-check-regression_part_2/index.html

Symmetric column changes only

-----------------------------

All other changes were because of failing existing unit tests (found via `mvn clean verify`).

I will let the build fail to show that the changes in this PR are dependent on the changes at https://github.com/checkstyle/checkstyle/pull/10104, then I will change the Checkstyle repo to my PR to show that all pass.

Failed build (I had to add `mvn clean verify` to CI):

```
Results :

Failed tests: 
  NoMainMethodInAbstractClassCheckTest.testDefault:87->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 0 expected:<...actClassCheck.java:1[00]:5: Main method insi...> but was:<...actClassCheck.java:1[16]:5: Main method insi...>
  ForbidWildcardAsReturnTypeCheckTest.testNewArrayStructure:461->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 0 expected:<[/pipeline/source/sevntu-checks/src/test/resources/com/github/sevntu/checkstyle/checks/design/InputForbidWildcardAsReturnTypeCheckNewArrayStructure.java:4:5: Wildcard as return type should be avoided].> but was:<[Audit done].>
  PreferMethodReferenceCheckTest.testObjectCreation:146->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 2 expected:<...ferenceCheck4.java:1[7]:39: Lambda can be r...> but was:<...ferenceCheck4.java:1[8]:39: Lambda can be r...>
  PreferMethodReferenceCheckTest.testStatementsListsWithExpressions:128->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 11 expected:<...eferenceCheck3.java:[46]:40: Lambda can be r...> but was:<...eferenceCheck3.java:[50]:40: Lambda can be r...>
  Jsr305AnnotationsCheckTest.testArrays:161->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 6 expected:<...kWithArrays.java:90:[5]: Return value must ...> but was:<...kWithArrays.java:90:[8]: Return value must ...>
  CustomDeclarationOrderCheckTest.mainMethod:189->AbstractModuleTestSupport.verify:229->AbstractModuleTestSupport.verify:267->AbstractModuleTestSupport.verify:304 error message 0 expected:<...d(private )' then 'M[ainM]ethod(.*)'.> but was:<...d(private )' then 'M[]ethod(.*)'.>

Tests run: 411, Failures: 6, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 32.994 s
[INFO] Finished at: 2021-06-16T11:42:15Z
[INFO] Final Memory: 56M/980M
[INFO] ---------------------------------------------------------------------

```
Full log at https://app.wercker.com/checkstyle/sevntu.checkstyle/runs/build/60c9e32c237ec900086455fa?step=60c9e3678ea84100087595c9

-------------

`mvn clean verify` is passing with Checkstyle version from  https://github.com/checkstyle/checkstyle/pull/10104

### CODE REVIEW

**Comment by rnveach** [comment]
> the changes in this PR are dependent on the changes at checkstyle/checkstyle#10104

This means this PR is dependent on 4 things before it can be merged:
1)  checkstyle/checkstyle#10104 must be merged
2) checkstyle must release those changes
3) eclipse-cs must release with that new version of checkstyle
4) The version of checkstyle in sevntu must update to the new eclipse-cs version (see https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/master/sevntu-checks/pom.xml#L18-L19 )

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/CustomDeclarationOrderCheck.java [hunk 1]
```diff
@@ -1145,10 +1145,8 @@ private static boolean hasOnlyStringArrayParameter(final DetailAST parametersAST
             final DetailAST parameterTypeAST = parameterDefinitionAST
                     .findFirstToken(TokenTypes.TYPE);
             if (hasChildToken(parameterTypeAST, TokenTypes.ARRAY_DECLARATOR)) {
-                final DetailAST arrayDeclaratorAST = parameterTypeAST
-                        .findFirstToken(TokenTypes.ARRAY_DECLARATOR);
                 final String parameterName =
-                        getIdentifier(arrayDeclaratorAST);
+                        parameterTypeAST.getFirstChild().getText();
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/CustomDeclarationOrderCheck.java:1149]: Parameter type is always first child now, it is no longer nested.

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java [hunk 2]
```diff
@@ -585,22 +586,17 @@ private static boolean isPrimitiveType(DetailAST ast) {
     /**
      * Checks whether token is array or elipsis.
      *
-     * @param identToken
+     * @param typeToken
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java:589]: Previously named `identToken`, this DetailAST was actually never an IDENT. In every UT we had, this method was passed a `TYPE` ast.

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java [hunk 3]
```diff
@@ -585,22 +586,17 @@ private static boolean isPrimitiveType(DetailAST ast) {
...
-                result = false;
-        }
-        return result;
+    private static boolean isArrayOrElipsis(final DetailAST typeToken) {
+        final DetailAST next = typeToken.getNextSibling();
+        final boolean isArrayDeclarator =
+                typeToken.findFirstToken(TokenTypes.ARRAY_DECLARATOR) != null;
+        final boolean hasArrayOrEllipses =
+                TokenUtil.isOfType(next, TokenTypes.ARRAY_DECLARATOR, TokenTypes.ELLIPSIS);
+        return hasArrayOrEllipses || isArrayDeclarator;
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java:599]: Switch was overkill for two branches, refactored this and modified logic a bit for new `type` ast.

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/MultipleVariableDeclarationsExtendedCheck.java [hunk 4]
```diff
@@ -167,9 +167,7 @@ private static DetailAST getLastNode(final DetailAST node) {
         while (child != null) {
             final DetailAST newNode = getLastNode(child);
             if (newNode.getLineNo() > currentNode.getLineNo()
-                    || newNode.getLineNo()
-                        == currentNode.getLineNo() && newNode
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/MultipleVariableDeclarationsExtendedCheck.java:171]: After main repo PR, this line was not covered, and no cases to cover it were discovered in regression testing.

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 5]
```diff
@@ -424,17 +424,22 @@ private static boolean isAnonClass(DetailAST ast) {
...
     private static boolean isArgUsedInArrayInit(DetailAST declarator, String name) {
-        final DetailAST nestedDecl = declarator.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
-        final DetailAST expr = declarator.findFirstToken(TokenTypes.EXPR);
-        boolean result = false;
-        if (nestedDecl == null && expr != null) {
-            final DetailAST ident = expr.findFirstToken(TokenTypes.IDENT);
-            result = ident != null && name.equals(ident.getText());
-        }
-        else if (nestedDecl != null && expr == null) {
-            result = isArgUsedInArrayInit(nestedDecl, name);
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheck.java:435]: No need for recursion here anymore, we can just iterate through the sequential `ARRAY_DECLARATOR` brackets and check if they have an `EXPR` child (i.e. `[2]`), then simply check the first `[a]` to make sure that the lambda parameter and expression match.

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/design/NoMainMethodInAbstractClassCheck.java [hunk 6]
```diff
@@ -223,10 +223,8 @@ private static boolean hasOnlyStringArrayParameter(final DetailAST parametersAST
             final DetailAST parameterTypeAST = parameterDefinitionAST
                     .findFirstToken(TokenTypes.TYPE);
             if (hasChildToken(parameterTypeAST, TokenTypes.ARRAY_DECLARATOR)) {
-                final DetailAST arrayDeclaratorAST = parameterTypeAST
-                        .findFirstToken(TokenTypes.ARRAY_DECLARATOR);
                 final String parameterName =
-                        getIdentifier(arrayDeclaratorAST);
+                        parameterTypeAST.getFirstChild().getText();
```
Comment (nmancus1) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/design/NoMainMethodInAbstractClassCheck.java:227]: Same as above, type is always the first child.

File: sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/ForbidWildcardAsReturnTypeCheckTest.java [hunk 7]
```diff
@@ -448,6 +448,21 @@ public final void testFullyQualifiedAnnotation()
...
     }
 
+    @Test
+    public final void testNewArrayStructure()
+            throws Exception {
+        final DefaultConfiguration checkConfig =
+                createModuleConfig(ForbidWildcardAsReturnTypeCheck.class);
+
+        final String[] expected = {
+            "4:5: " + getCheckMessage(MSG_KEY),
```
Comment (nmancus1) [inline sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/design/ForbidWildcardAsReturnTypeCheckTest.java:458]: New violation found in regression testing.

File: wercker.yml [hunk 8]
```diff
@@ -54,6 +54,23 @@ build:
...
+      code: |
+        if [[ $SKIP_CI == 'false' ]]; then
+          cd sevntu-checks
+          mkdir -p .ci-temp
+          cd .ci-temp
+          git clone -b issue-10100 https://github.com/nmancus1/checkstyle.git
+          cd checkstyle
+          mvn -e clean install -Pno-validations
+          cd ../..
+          mvn clean verify -Dcheckstyle.eclipse-cs.version=8.44-SNAPSHOT
```
Comment (nmancus1) [inline wercker.yml:68]: Temporary hack to show that we can pass `mvn clean verify` with checkstyle version from base repo PR.

**Comment by rnveach** [comment]
Added blocked label because of https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/851#issuecomment-861974113 to prevent accidental merging

**Comment by kaushal-khurana-17** [comment]
Hi, I am a beginner at this issue can you help me. How to start it?

**Comment by kaushal-khurana-17** [comment]
Hi, can you help me?

**Pull request merged**